### DISCUSSION TITLE
using unit model (heater as a boiler or a condenser)

### DISCUSSION THREAD

**Discussion by Yazeede73** [IDAES/idaes-pse#533] [Q&A]
I have been trying to use the heater generic unit model as a boiler and a condenser while trying to simulate a basic Rankine cycle flowsheet. however the vapor fractions never changes further more solution keeps converging to an infeasible point. is there any way to use the heater unit model as a boiler or a condenser anyway? 

Thanks for your help in advance.

**Comment by andrewlee94** [comment]
@Yazeede73 We will need more information about what you are trying to do in order to help - e.g.  what property package are you using and what are in inlet conditions to the heater? Also, what solver ae you using, and what is the full output from the solver at the end of the run?

The fact that you are getting an infeasible solution in itself is telling - it implies that there is no solution that satisfies the conditions you are trying to model. In many cases, this is due to bounds placed on one or more of the variable in the problem, but without more details that is all we can tell you. This is also why the vapour fraction never changes - the solver cannot find a new solution to the problem, so there is no new value for vapour fraction.

**Reply by Yazeede73** [reply]
ok so i am trying to use IDAES for my research work. but before that i wanted to make sure i can use it correctly so i brought out a simple Rankine Cycle steam turbine problem which consists of 4 Units (boiler, Turbine, condenser and pump) the property package i am using is IAPWS95.  as for the solver i am following mostly the workshops that you have over on youtube so it is Ipopt and here is the end result of the solution. 

"EXIT: Converged to a point of local infeasibility. Problem may be infeasible.
WARNING: Loading a SolverResults object with a warning status into
    model.name="unknown";
      - termination condition: infeasible
      - message from solver: Ipopt 3.13.2\x3a Converged to a locally
        infeasible point. Problem may be infeasible."

as for the conditions(inputs) i put in are as follows;
m.fs.boiler.inlet.pressure[0].fix(1980000)
m.fs.boiler.outlet.enth_mol[0].fix((value(iapws95.htpx(T=773*units.K, P=1980000*units.Pa))))
m.fs.turbine.outlet.pressure[0].fix(10000)
m.fs.pump.outlet.flow_mol[0].fix(55.5)
m.fs.pump.outlet.pressure[0].fix(1980000)
m.fs.turbine.inlet.pressure[0].fix(1980000)

and the arcs are shown below;
m.fs.s01=Arc(source=m.fs.boiler.outlet, destination=m.fs.turbine.inlet)
m.fs.s02=Arc(source=m.fs.turbine.outlet, destination=m.fs.condensor.inlet)
m.fs.s03=Arc(source=m.fs.condensor.outlet, destination=m.fs.pump.inlet)
m.fs.s04=Arc(source=m.fs.pump.outlet, destination=m.fs.boiler.inlet)

i hope these details are enough. Thanks a lot for your help in advance it really helps my understanding.

**Comment by Andresj89** [comment]
@Yazeede73 you can certainly use a generic heater to represent a boiler or condenser. It is very important to initialize every unit, I believe for the generic heater you need to specify inlet (flow, enthalpy, pressure if you are using IAPWS95 as property package) and heat duty (just an estimate) so the initialization calculates the outlet. Make sure to initialize the turbine, and pump too. Then you can solve the entire flowsheet making sure that `degrees_of_freedom(m) == 0`. If you don know the enthalpy and you are using IAPWS95 then you can import the function `htpx()` to calculate enthalpy based on temperature and pressure, you should use it in combination with Pyomo units as follows: 
```
from idaes.generic_models.properties.iapws95 import htpx
from pyomo.environ import units
m.fs.boiler.inlet.enth_mol.fix(htpx(T=value*units*K, P=value*units*Pa))
```
To initialize simply do:
`m.fs.boiler.initialize()`

If you want to know the current values, do:
`m.fs.boiler.report()`

Hope this helps.

**Reply by Yazeede73** [reply]
@Andresj89 thanks for your reply......is it necessary to estimate the heat duty?...what if i fix the boilers outlet temperature. because i set that the generic heater has no pressure change so i assumed i would get the heat duty as a result by inputting the inlet conditions you mentioned and the outlet temperature. furthermore, how can assert phase change in the process, i tried looking up how on the documentation but could not find anything that would help. 

Thanks in advance

**Reply by Andresj89** [reply]
@Yazeede73 since the models we deal with in process synthesis are usually complex (that is, non-convex NLPs), it is highly recommended that each unit model is initialized before actually solving the model. In fact, for unit models such as heat exchangers, it is a must to initialize before solving it. We have initialization routines tailored to each unit model. In the case of the boiler (heater), you can see how to initialize it here: https://idaes-pse.readthedocs.io/en/stable/technical_specs/model_libraries/generic/unit_models/heater.html?highlight=heater. Once you initialize it you can do
```
model.fs.heater.heat_duty[:].unfix()
model.fs.heater.outlet[:].enth_mol.fix(htpx(T=OutletTemperature*units.K, P=OutletPressure*units.Pa))
```
And now you can solve the heater for heat duty given the outlet temperature. You can check the heater state by doing `model.fs.heater.report()`. You can also access the vapor fraction by doing: `model.fs.heater.control_volume.properties_in[0].vapor_frac.expr()` and `model.fs.heater.control_volume.properties_out[0].vapor_frac.expr()`

**Reply by Yazeede73** [reply]
ok but from what i learned i initialize in a sequential manner using "seq" function however i can try initializing each on its own...Furthermore how can i initialize the units if i do not have the all inlet conditions such as heat duty...For example i have inlet; flow, pressure and temperature and outlet; temperature.. how can i initialize a problem of this sort. also can i fix inlet vapor fraction and exit? lines you showed me above?
